processing/combine_factRu2016.py

```diff
@@ -0,0 +1,315 @@
+
+    # Following the for loop, `label_dict` is now a dict with span_ids as keys
+    # and their labels as values.
+    label_dict = {}
+    for d in ind_dicts:
+        label_dict.update(d)
+
+    return label_dict
+
+
+def create_span_df(extracted_span_data, label_dict):
+
+    """
+    Takes the extracted span data from 2 steps back (list of lists), and
+    the label dict from the last step, and creates the span + label DF for
+    joining with the token DF."""
+
+    # Create span DF from the extracted span data
+    span_df = pd.DataFrame(extracted_span_data).iloc[:, 1:]
+    span_df.columns = ['position_id', 'mention_type', 'begin_index', 'n_tokens']
+
+    # Convert dtypes for joining with token df
+    for col in ['position_id', 'n_tokens', 'begin_index']:
+        span_df[col] = span_df[col].astype('int')
+
+    # Assign labels to each position
+    get_label = lambda x: label_dict.get(str(x), np.nan)
+    span_df['label'] = span_df['position_id'].apply(get_label)
+
+    return span_df
+
+
+def create_final_label_df(token_df, span_df):
+
+    """
+    create_final_label_df joins the token_df and span_df together,
+    then applies nested conditional logic to assign the correct prefix
+    to each label.
+
+    The possible situations:
+      1. (cond1) A single token has a label and the word before it is not labeled
+      2. (cond2) A token is labeled, and is the first word of a multi-word phrase
+      3. (cond3) A token is labeled, and is a non-first word of a multi-word phrase
+
+    (1) and (2) correspond to a 'B-' prefix, and (3) corresponds to the 'I-' prefix.
+    These are standard prefixes used in NER to help the model know which tokens
+    are the "B"eginning of a labeled phrase and which are "I"nside the phrase.
+
+    This ends up looking like B-ORG (organization), I-PER (person), etc.
+    The four possible labels in this dataset are Org, Location, Person, and LocOrg.
+    """
+
+    label_df = token_df.merge(span_df,
+                              left_on='begin_index',
+                              right_on='begin_index',
+                              how='left')
+
+    # Position IDs being associated with multiple mention types created
+    # duplicate tokens in the join, so we drop those now
+    label_df.drop_duplicates(subset=['begin_index', 'text'],
+                             keep='first', inplace=True)
+
+    # SQL-like way of keeping track of the n_tokens value from prev row
+    label_df['lag(n_tokens)'] = label_df['n_tokens'].shift(1)
+
+    cond1 = (label_df['n_tokens'] == 1) & (label_df['lag(n_tokens)'].isna())
+    label_df['label1'] = np.where(cond1, 'B-' + label_df['label'], np.nan)
+
+    cond2 = (label_df['n_tokens'] >= 1) & (label_df['lag(n_tokens)'].isna())
+    label_df['label2'] = np.where(cond2, 'B-' + label_df['label'], np.nan)
+
+    # Label and lag(n_tokens) are both non-null (indicates being inside of a phrase)
+    cond3 = (label_df['label'].notna()) & (label_df['lag(n_tokens)'].notna())
```

Comment:
Is there output for how many are in each condition?

Reply:
clarification: how many rows fall into cond1, cond2, cond3